pkg/runner/local_docker.go

```diff
@@ -195,81 +205,116 @@ func (r *LocalDockerRunner) setupSyncClient() error {
+	// We have to pass its group id to the container, so that it can send us back messages
+	// with its own id.
+	expectingOutcomes := result.countTotalInstances()
 	done := make(chan bool)
 
 	go func() {
 		running := true
-		for running {
+		for running && expectingOutcomes > 0 {
 			select {
 			case <-ctx.Done():
 				running = false
 			case e := <-eventsCh:
-				// for now we emit only outcome OK events, so no need for more checks
 				if e.SuccessEvent != nil {
-					se := e.SuccessEvent
-					o := result.Outcomes[se.TestGroupID]
-					o.Ok = o.Ok + 1
+					result.addOutcome(e.SuccessEvent.TestGroupID, task.OutcomeSuccess)
+					expectingOutcomes -= 1
+				} else if e.FailureEvent != nil {
+					result.addOutcome(e.FailureEvent.TestGroupID, task.OutcomeFailure)
+					expectingOutcomes -= 1
+				} else if e.CrashEvent != nil {
+					result.addOutcome(e.CrashEvent.TestGroupID, task.OutcomeFailure)
+					expectingOutcomes -= 1
 				}
+				// else: skip
 			}
 		}
 
-		result.Outcome = task.OutcomeSuccess
-		if len(result.Outcomes) == 0 {
-			result.Outcome = task.OutcomeFailure
-		}
-
-		for g := range result.Outcomes {
-			if result.Outcomes[g].Total != result.Outcomes[g].Ok {
-				result.Outcome = task.OutcomeFailure
-				break
-			}
-		}
-
+		result.updateOutcome()
 		done <- true
 	}()
 
 	return done, nil
 }
 
-func (r *LocalDockerRunner) Run(ctx context.Context, input *api.RunInput, ow *rpc.OutputWriter) (runoutput *api.RunOutput, err error) {
-	log := ow.With("runner", "local:docker", "run_id", input.RunID)
+func (r *LocalDockerRunner) prepareOutputDirectory(instance_id int, runenv *runtime.RunParams) (string, error) {
+	// <outputs_dir>/<plan>/<run_id>/<group_id>/<instance_number>
+	odir := filepath.Join(r.outputsDir, runenv.TestPlan, runenv.TestRun, runenv.TestGroupID, strconv.Itoa(instance_id))
 
-	err = r.setupSyncClient()
+	err := os.MkdirAll(odir, 0777)
 	if err != nil {
-		log.Error(err)
-		return
+		return "", fmt.Errorf("failed to create outputs dir %s: %w", odir, err)
 	}
 
-	// Grab a read lock. This will allow many runs to run simultaneously, but
-	// they will be exclusive of state-altering healthchecks.
-	r.lk.RLock()
-	defer r.lk.RUnlock()
+	return odir, nil
+}
+
+func (r *LocalDockerRunner) prepareTemporaryDirectory(instance_id int, runenv *runtime.RunParams) (string, error) {
+	var tmpdir string
+	tmpdir, err := ioutil.TempDir("", "testground")
+	if err != nil {
+		return "", fmt.Errorf("failed to create temp dir: %s: %w", tmpdir, err)
+	}
+
+	return tmpdir, nil
+}
 
-	result := newResult()
+func (r *LocalDockerRunner) Run(ctx context.Context, input *api.RunInput, ow *rpc.OutputWriter) (runoutput *api.RunOutput, err error) {
+	log := ow.With("runner", "local:docker", "run_id", input.RunID)
+
+	result := newResult(input)
 	runoutput = &api.RunOutput{
 		RunID:  input.RunID,
 		Result: result,
 	}
 
 	defer func() {
+		if err != nil && result.Outcome == "" {
+			result.Outcome = task.OutcomeFailure
+		}
```

Comment:
Is it possible to have an error, but the outcome is not an empty string? I would assume that the outcome should be a failure, even in such a case.

Reply:
From what I gathered `result.Outcome == ""` indicates the outcome wasn't previously set.

Reply:
Exactly, it's a catch-all "if we exit with an error, and the Outcome was not set, then always fall back to failure"